nh_clinical/patient.py

```diff
@@ -51,7 +51,34 @@ class NhClinicalPatient(osv.Model):
     ]
 
     @staticmethod
-    def _clean_identifiers(dirty_vals):
+    def _check_identifier_for_bad_chars(value):
+        """
+        Check for bad characters in string
+
+        :param value: string to check
+        """
+        unallowed_chars = '[^a-zA-Z0-9_\-\s]'
+        allowed_chars = re.sub(unallowed_chars, '', value)
+        if allowed_chars != value:
+            raise ValidationError(
+                'Patient identifier can only contain '
+                'alphanumeric characters, hyphens and underscores'
+            )
```

Comment:
Couple of points here:

1. This regex allows trailing spaces which are meant to be invalid. I know that we're being nice by allowing the user to enter spaces that are technically invalid, but why not remove whitespace first so that we don't have to include them in the regex here? It waters down the meaning of the method a bit to be checking for bad chars but still allow spaces.

2. If I understand correctly the logic is to replace every invalid character with nothing and then check if the value has changed. The logic can be more straightforward if you just search for invalid characters which I think you can do with:

```python
allowed_chars = '[a-zA-Z0-9_\-]+'
match = re.match(allowed_chars, value)
if not match:
    raise ValidationError("bork.")
```